File: backend/app/schemas/schedule.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, Literal
from datetime import datetime


class IntervalConfig(BaseModel):
    value: int
    unit: Literal['minutes', 'hours', 'days']


class DailyConfig(BaseModel):
    time: str  # "HH:MM" format


class WeeklyConfig(BaseModel):
    days: list[Literal['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']]
    time: str  # "HH:MM" format


class ScheduleCreate(BaseModel):
    mapping_id: str
    mode: Literal['once', 'scheduled']
    schedule_type: Optional[Literal['interval', 'daily', 'weekly', 'cron']] = None
    interval: Optional[IntervalConfig] = None
    daily: Optional[DailyConfig] = None
    weekly: Optional[WeeklyConfig] = None
    cron_expression: Optional[str] = None
    enabled: bool = True


class ScheduleResponse(ScheduleCreate):
    id: str = Field(alias='_id')
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(populate_by_name=True)

File: ui/src/types/schedule.ts
export type ScheduleMode = 'once' | 'scheduled';
export type ScheduleType = 'interval' | 'daily' | 'weekly' | 'cron';
export type IntervalUnit = 'minutes' | 'hours' | 'days';
export type WeekDay = 'mon' | 'tue' | 'wed' | 'thu' | 'fri' | 'sat' | 'sun';

export interface IntervalConfig {
  value: number;
  unit: IntervalUnit;
}

export interface DailyConfig {
  time: string; // "HH:MM"
}

export interface WeeklyConfig {
  days: WeekDay[];
  time: string; // "HH:MM"
}

export interface ScheduleConfiguration {
  mapping_id: string;
  mode: ScheduleMode;
  schedule_type?: ScheduleType;
  interval?: IntervalConfig;
  daily?: DailyConfig;
  weekly?: WeeklyConfig;
  cron_expression?: string;
  enabled: boolean;
}

export interface ScheduleResponse extends ScheduleConfiguration {
  _id: string;
  created_at: string;
  updated_at: string;
}
